### ISSUE TITLE
Old Issues From SourceForge?

### ISSUE THREAD

**Issue by jfelchner** [tmux/tmux#34]
There was a very long thread about an issue that I was interested in on SourceForge.  I'd be happy to convert it over, but it seems as though the link is no longer valid.

Does one of the maintainers still have access to the old SourceForge issues?

**Comment by jfelchner** [comment]
It was specifically issue number 140 which was the issue about truecolor support in Tmux.

The old URL was: http://sourceforge.net/p/tmux/tickets/140/

**Comment by jfelchner** [comment]
@ThomasAdam what is going on?

**Comment by jfelchner** [comment]
@ThomasAdam why did you close this?

**Comment by ThomasAdam** [comment]
Hmm.  There was supposed to be a comment from me.  Not sure what happened there.

We had a patch for this in the past.  I'm unsure how useful 24bit colour support is, it's certainly not something happening by default, and only a handful of applications support this.  Until such time it's more ubiquitous, this isn't worth it.

**Comment by jfelchner** [comment]
@ThomasAdam that's everything that was in that ticket.  Evidence that it is currently _very_ ubiquitous.  Almost every major terminal currently supports it.  That ticket had links to tickets on all of them outlining their support for it.

**Comment by jfelchner** [comment]
I'd rather not retype all of that if someone can get access to the old ticket.

**Comment by ThomasAdam** [comment]
You mean this:  https://www.mail-archive.com/tmux-users@lists.sourceforge.net/msg06296.html

**Comment by jfelchner** [comment]
That's the first part to it yes. :grinning: thanks!

**Comment by ThomasAdam** [comment]
There never was a follow-up.  But it shows that libvte supports it, and almost no applications.  What XTerm does, is what tmux does now with 256 colour support, in terms of choosing the nearest colour match.

**Comment by jfelchner** [comment]
@ThomasAdam it may have been a different issue then that had all the followups.  There were a lot including some from Nicholas.  Is there a way for me to search the archives?

**Comment by jfelchner** [comment]
Here's what iTerm does:

![image](https://cloud.githubusercontent.com/assets/285582/8394546/76a1556e-1d02-11e5-8cc7-3c59a54a1d29.png)

Here's what Tmux does:

![image](https://cloud.githubusercontent.com/assets/285582/8394548/8512cfce-1d02-11e5-9bdd-a3cf9b8bcf49.png)

**Comment by ThomasAdam** [comment]
The issue you referenced is what I've linked you to.  Yes, the archives are available in www.mail-archive.com -- you can see that from where I've linked you to previous posts.  Go forth and search

**Comment by jfelchner** [comment]
@ThomasAdam here's the patch: https://gist.github.com/JohnMorales/0579990993f6dec19e83

Seems pretty small to get such a great win. I'd like to be able to use the full color range in terminal.  Emacs and Neovim are going to be releasing support for truecolor soon.  Neovim already has it in `master` and Emacs is getting it in the next release.

**Comment by ThomasAdam** [comment]
Ugh, that would need a little bit of work.  Either way, this has already been discussed.

**Comment by jfelchner** [comment]
@ThomasAdam I know it has, there was a back and forth with Nicholas about it and he seemed receptive to it.  Unfortunately I can't find that discussion in the archives for some reason.  But they definitely happened.My viewpoint is that if everyone waits for everyone else to support it then no one is ever going to support it.

**Comment by jfelchner** [comment]
@ThomasAdam I do understand that xterm doesn't support truecolor.  And I also understand that the terminal you're using must also support truecolor.  All I'm saying is that a terminal can support truecolor even if xterm doesn't.  And the list of terminals which _don't_ support truecolor is effectively down to two:

* urxvt - http://lists.schmorp.de/pipermail/rxvt-unicode/2013q3/001826.html
* Terminology (E17) - https://phab.enlightenment.org/T746

All of the other major terminals support it and they support it in the same standardized format.

* So, is it true that xterm doesn't support truecolor?  Yes.
* Is it true that despite that, nearly every popular terminal supports it anyway? Yes.
* So do I think that it's worthwhile for tmux to support it so that the users of those terminals can use Tmux with native truecolor support?  Absolutely.

**Comment by nicm** [comment]
I do not want issues that just sit open with no progress like that SF issue. Support for this as been discussed already and nobody has provided code and been willing to move it forwards, the way I added it to tmux is all that is going to happen until someone does.

**Comment by jfelchner** [comment]
@nicm that's completely understandable.  I just wanted to make sure that the SF conversation wasn't lost and that there was a place for people to comment and or reference when a PR was submitted.

Like I said, I'd do this myself in a heartbeat if I knew any C. :)

Thanks again!

**Comment by willghatch** [comment]
Hey, I'd like to +1 this feature as well.  There is a patch floating around here: http://hoppsjots.org/?p=172 that I've used before (I didn't really check its quality much, and it's been a long time since I've used it, but it worked well then).  I've been using various patches to support true color in tmux and emacs for over a year on and off (emacs more so, as I use tmux less often).  Aside from emacs I haven't used these patches in a while due to the hassle of having to patch and build everything from source, so I really want this feature mainlined.  Most terminal emulators support true color now, in particular kde console and gnome terminal which have for a long time.  Most people I know use one of those two, and most of the rest still use one that supports 24-bit color.  so there is a large user base that would benefit from this.

**Comment by XVilka** [comment]
@nicm @jfelchner well, there is a patch available https://gist.github.com/JohnMorales/0579990993f6dec19e83

**Comment by ThomasAdam** [comment]
@XVilka - In reference to the above, I've already noted it needs cleaning up.  I'll try and take a look at doing so this weekend, but no promises at all.

**Comment by jfelchner** [comment]
@ThomasAdam you're awesome! :)  I understand your time is valuable so I appreciate anything you can do.  Thanks again.

**Comment by cgag** [comment]
Just here with another +1, it's a big enough win for me that I left urxvt so I could have true color support.

**Comment by eidge** [comment]
+1

**Comment by dmerc** [comment]
If this changes anything, here is another user that would like to see true color in tmux. +1

**Comment by mindshell** [comment]
+1

**Comment by jfelchner** [comment]
To anyone who comes after, the +1's are really the only way to show interest in an issue on github, but they lack something.

Maybe try putting the reason why you would like the feature. Maybe "I would love to be able to use any color I want in my Vim themes".  Doing so may bring to light that there are use cases that others haven't thought about.

**Comment by mindshell** [comment]
@jfelchner Using neovim as replacement for vim and would like to use neovim in tmux using true colors.

**Comment by dueyfinster** [comment]
@jfelchner I have true color in gnome-terminal and immediately lose it going to tmux. Makes me sad, as I'm a big believer in keyboard productivity and want to use tmux everywhere. More and more terminals are supporting it. Just waiting on tmux ! +1 for true color!

**Comment by XVilka** [comment]
I'm using both radare2 and neovim, both supporting truecolors, currently I've forced to use them outside the tmux session because of support. I loved urxvt before, but now switched to the mlterm (libvte-based), because of 24bit color and correct RTL support.

**Comment by zsoltf** [comment]
+1

Three main things I can't live without:
- Vim with the true color patch. It's been amazing to preview html colors and use plugins like ColorV.
- elinks in true color mode to preview css colors.
- Image previews over ssh.

**Comment by pickfire** [comment]
Terminals that support true colors. https://gist.github.com/XVilka/8346728

**Comment by milabs** [comment]
+100500

**Comment by korin** [comment]
The same here, I have terminal and editor with true color support but neovim looks awful in tmux.
I understand that right now it's hard to detect true color support but maybe tmux can use the same approach like neovim? Just env flag so I can tell tmux that true color is on. 
@ThomasAdam what you think?

**Comment by mhartington** [comment]
+1 for this. Like the idea as env flag. Or even add a build option. Build option would makes sense with package managers in mind, like homebrew.

**Comment by ThomasAdam** [comment]
@korin - I think there's been far, far too much idle chatter about this.  We *understand* people are desperate to jump onto the nice coloured 24-bit bandwagon and rejoice in using those minor handful of programs supporting it.

But it won't make me look at this any faster, alas, my time is very scarce at the moment.

And no, an environment variable is not the correct thing to do at all for this.

So please, I am getting a little bored of getting notifications for +1s.  This will get done at some point, I'm sure.

**Comment by jfelchner** [comment]
@ThomasAdam I think it's important to have the discussion and for people to give their reasoning behind why they want it.  Maybe there's something that we haven't thought of?  I also think it's important to show how much interest is behind having this be implemented.  After all, if 0 people wanted it, then there's no reason to do it.  If thousands want it, well, that gives it a little more weight.

I agree with you that I really do wish that Github had an option to +1 without actually adding a new comment however.  As I said above, a "+1" with no substance is just extra noise.

**Comment by sunaku** [comment]
[The subscribe button](https://help.github.com/articles/subscribing-to-conversations/) is the closest thing to a silent +1 that GitHub currently seems to offer. :speak_no_evil:

**Comment by XVilka** [comment]
@ThomasAdam Here is the updated version of the patch https://gist.github.com/zchee/9f6f2ca17acf49e04088
![screenshot](https://cloud.githubusercontent.com/assets/6366270/9142997/613db37e-3d7e-11e5-8984-c7d4137eb63c.png)

Please review if you have some time. Looks clean for me.

**Comment by nicm** [comment]
Are you serious with this patch? Why do you revert half of colour.c and the 24 bit colour parts of input.c?

It is not acceptable to permit : as well as ; for everything, it would need to accept : just for the 24 bit escape sequences, although I suggest not accepting them at all and hopefully they will die out.

I do not want to have 24 bit colour support as well as 256 colour. It needs to supersede 256 colour (and emulate it instead).

Also we will need a way to tell if the terminal supports 24 bit colour. I suggest adding a Tc terminfo capability for now.

Please don't include commented out logging or // comments.

**Comment by XVilka** [comment]
@nicm this is not my patch, but why not accept it after some cleanup?

About 24bit color in terminfo - yes, this is what is need to be done, since:
> currently there is no way to indicate 24 bit color support via the mainstream termcap/terminfo.
> See the corresponding thread/messages from the ncurses maintainer
> https://lists.gnu.org/archive/html/bug-ncurses/2013-10/msg00007.html

And official terminfo maintainer strongly opposing the 24bit colors :(

**Comment by nicm** [comment]
If you are going to cleanup the patch then please do so.

**Comment by jfelchner** [comment]
@XVilka is there a way to get the person who maintains that patch to join the conversation here?

**Comment by XVilka** [comment]
@jfelchner of course, just cast him: @zchee

**Comment by nicm** [comment]
It is also important not to forget about grid_string_cells in grid.c.

**Comment by jfelchner** [comment]
@zchee do you have any interest in helping here? I have a feeling @nicm is going to want things done a very specific way but once it's in, no one will have to do patches anymore and everyone can benefit.

**Comment by borring** [comment]
@ThomasAdam It seems like you're saying that TrueColor support is not very important because only a small quantity of applications support it.

I'm not trying to get you to work faster. As you said, the work is probably going to be done sooner or later. But I don't want you to feel like you're putting time into unimportant features. IMO it's not the number of applications that is important. It's the time spent using those applications. For example, most people probably don't spend hours looking at `top` output. On the other hand, people _do_ spend hours in emacs and vim/neovim and the like.

So this feature _does_ have a significant impact on usability despite having only a few support applications.

**Comment by dequis** [comment]
At this point, I don't think we need to convince the tmux devs with words, we just need to clean that patch. (I could give it a shot, but i'm far from an expert with terminal black magic)

**Comment by zchee** [comment]
@jfelchner Sorry, I can't help for you.
I do not know exactly the C language. I first wrote C languege is that patch.
I do that patch wrote in intuition.

Sorry.

**Comment by jfelchner** [comment]
@zchee thanks for replying.  I appreciate it. :smile:

**Comment by pickfire** [comment]
@zchee taking the first step to create a c language patch is not as easy as abc. Good job in doing that, I just hope that the tmux developers will accept it.

**Comment by jfelchner** [comment]
@pickfire there's almost zero chance of that. They obviously care about the code base and that patch is not up to their standards. We need to find a champion who will take this on.

**Comment by pickfire** [comment]
Well, I knew **some** c language, I am still learning it. I think I will take a look in it.

**Comment by zchee** [comment]
@pickfire Yes,yes. I know.
but, It would not be completed unless the challenge? :)

However, I also hope that me rewrite someone who has excelled in the c language.

**Comment by pickfire** [comment]
I had look at the code, I doesn't understand much but I think you can copy some code from other applications that have true color support.

**Comment by sunaku** [comment]
Here's my attempt at simplifying @zchee's patch in PR #112. :smiley_cat:

**Comment by zchee** [comment]
@sunaku Thanks :smile:

**Comment by saitoha** [comment]
@jfelchner Do you know this specification has obvious backward compatible problem?
You need to reserve another new private sequence. Don't extend SGR with wrong way that lacks backward compatibility.
I think this specification comes from wrong interpretation of ECMA-48 and CCITT Recommendation T.416.
It is important problem for me and some old software, although It might be a trivial issue for you.
If you can't do it for some reason, at least you need to define new pair of query and response sequences that indicates whether the terminal support it. Please don't think that terminfo and $TERM solves all problems.

**Comment by dequis** [comment]
>Please don't think that terminfo and $TERM solves all problems.

Why not? Seems to me that terminfo makes it completely opt-in, but enlighten me if that's wrong.

**Comment by XVilka** [comment]
@saitoha @dequis current terminfo doesn't have support for the 24bit colors - see the general usage example of them at this comment https://gist.github.com/XVilka/8346728#gistcomment-1491237
But here is the problem: currently there is no way to indicate 24 bit color support via the mainstream termcap/terminfo. See the corresponding thread/messages from the ncurses maintainer https://lists.gnu.org/archive/html/bug-ncurses/2013-10/msg00007.html
The only way is to fork terminfo/termcap or ask somehow to push the extension.

**Comment by jfelchner** [comment]
Can we specify a new TERM string like `screen-16bit` and check that? Then the user can manually opt in. Or maybe add a `TERM_COLOR_DEPTH` environment variable?

**Comment by saitoha** [comment]
@XVilka Yeah, I have also read that threads of ncurses in real-time.
I think the situation would not be improved if terminfo should support it or ncurses is forked.
I think we no longer should not rely incomplete mechanisms such as terminfo. Especially in this case, it is not available over ssh to existent old systems.

To introduce 24bit color without causing confusion, we should do the following two things.

1. Define new pair of query and response sequences that indicates whether the terminal support 24bit color. This mechanism is needed for certain negotiation between terminals and applications.

2. Revoke current extended-SGR format and define new 24bit color sequence. This work will prevent to cause problems on some terminal filter programs, and keep compatibility against thousands of existent terminals.

**Comment by dequis** [comment]
In #112, @nicm suggested checking for "Tc" in terminfo, which is something that is not used anywhere yet, but can be added manually with terminal-overrides in the tmux config. That might be a better way to convince the terminfo maintainer, but even if that doesn't happen, it's an opt-in method.

The use case of "ssh to existent old systems" can be covered through terminal-overrides too, if you wish. You'd have that problem with any kind of newer control sequences anyway, even 256 colors. 

Using query/response sequences overcomplicates things, and is not even 100% safe. To give an example of an existing feature that works that way, there's bracketed paste, which applications tend to just have as a configuration setting, because there's no guarantee that all terminals will handle the query transparently. Readline has the most "clever" way to implement bracketed paste, it looks at terminfo for the existence of similar escape sequences and uses that to do a hopefully accurate guess that it's safe to send a query.

Let's not do that. Terminfo is saner.

**Comment by saitoha** [comment]
> You'd have that problem with any kind of newer control sequences anyway, even 256 colors. 
Yes, I had many troubles for Xterm's 256 color sequence (It wasn't also designed safely) for this 10 years. It had takes many years for terminfo covers it. I don't want to repeat it.
On the other hand, many other new control sequences such as bracketed paste seems to be designed safely. For example, To enable bracketed paste mode, mode switch (CSI ?2004h) is required.

**Comment by 6D65** [comment]
+1 It would definitely be great to have tmux and neovim. currently tmux messes up even the 256 colors schemes.

**Comment by dani-h** [comment]
Are there any plans on merging this to master in the near future?

**Comment by ThomasAdam** [comment]
@dani-h -- I appreciate there's a lot of bullshit comments from people in this thread, but the most recent  one clearly says there's more work to be done, and until the patches are reviewed further, it's not being merged.

**Comment by lvht** [comment]
:+1: using NeoVim + st + tmux

**Comment by metincc** [comment]
+1 I'm looking forward to this feature.

**Comment by sunaku** [comment]
Folks, I'm happy to tell you that this feature now exists in the master branch at 427b8204268af5548d09b830e101c59daa095df9! :gift: 

See pull request #112 for implementation details and read its [usage instructions here](http://sunaku.github.io/tmux-24bit-color.html#usage). :memo:

**Comment by pacuna** [comment]
anybody being able to do this with $TERM = xterm-256color ?
I'm following the instructions given by @sunaku but then when I execute `tmux info | grep Tc` I get nothing. I'm using version 2.1

Thanks.

**Comment by renstrom** [comment]
> I'm using version 2.1

@pacuna it's not available in v2.1, it's coming in the next release of tmux (2.3?). So you'll have to compile it yourself if you want to try it now.

**Comment by ThomasAdam** [comment]
It's in version 2,2; released already.

**Comment by pacuna** [comment]
thanks guys, I'll install the latest version and see how it goes

**Comment by deathmaz** [comment]
This is really amazing, i didn't believe my eyes first time i tried this feature, thank you very much !!!

**Comment by gfairchild** [comment]
I have a question on how to enable this. I can get this working with the following lines in my `.tmux.conf`:

```
set -g default-terminal "xterm-256color"
set-option -ga terminal-overrides ",xterm-256color:Tc"
```

I was previously using `screen-256color`. I've started to notice goofy issues here and there. For example, my home/end keys aren't working properly. Reading around, I'm seeing suggestions advising against `xterm-256color` (https://github.com/neovim/neovim/issues/2028, for example). Is it possible to get 24-bit colors working with `screen-256color`? I've tried `set-option -ga terminal-overrides ",screen-256color:Tc"`, but it doesn't seem to work.

**Comment by mikedfunk** [comment]
I just did this:
```
set -g default-terminal "screen-256color"
set-option -ga terminal-overrides ",screen-256color:Tc"
```
and it works

**Comment by sunaku** [comment]
@gfairchild Try removing the `default-terminal` setting from your tmux config:

    set -g default-terminal "xterm-256color"

And just the `terminal-overrides` to reflect your actual `$TERM` outside of tmux:

    set-option -ga terminal-overrides ",xterm-256color:Tc"

**Comment by gfairchild** [comment]
It's not working for me. Any ideas why? I'm using the latest iTerm2 3.0.7 and tmux 2.2.

**Comment by zchee** [comment]
@gfairchild What are you set `$TERM`?

**Comment by gfairchild** [comment]
Ah! That did it. Removing the `default-terminal` line and leaving in the `terminal-overrides` seems to have done it. Now, my `$TERM` is screen.

**Comment by zchee** [comment]
@gfairchild Oh? Sorry, still not work?
If so, try `export TERM=xterm-256color`

**Comment by gfairchild** [comment]
It seems to be working now. Thanks!

**Comment by yifanchen** [comment]
I am on Neovim, trying to get Atom color scheme working. It works completely fine after `let $NVIM_TUI_ENABLE_TRUE_COLOR=1` in `.vimrc`. However, it doesn't work right in Vim within Tmux. 
I figured that it might be some true color setting needed in my `.tmux.config`. So, I tried the following cmd that I saw on this thread, didn't seem working. Any ideas? 

```
set -g default-terminal "xterm-256color"
set-option -ga terminal-overrides ",xterm-256color:Tc"
```

```
set -g default-terminal "screen-256color"
set-option -ga terminal-overrides ",screen-256color:Tc"
```

I am on Tmux 2.1

**Comment by dequis** [comment]
@yifanchen please open a new ticket instead, posting comments here sends pointless notifications to more than 30 people

**Comment by yifanchen** [comment]
@dequis Saw the comments early, didn't want to make a duplicated ticket. Since now you said it, will do then. Thanks.

**Comment by blark** [comment]
Any info on how to make this work if you're using multiple terminals? Sometimes I SSH into the box using urxvt which by default has a $TERM of `rxvt-unicode-256color` and other times I use my Mac and iTerm2 which has it's terminal set to `xterm-256color`. Currently truecolor only works on the mac as I've put `set-option -ga terminal-overrides ",xterm-256color:Tc"` in my config file. 

I've tried to do something like `set-option -ga terminal-overrides ",$TERM:Tc"` but it doesn't seem to work.

**Comment by jwhitley** [comment]
@blark, from the tmux man page (v2.3):

> **terminal-overrides string**
> 
> Contains a list of entries which override terminal descriptions read using `terminfo(5)`.  string is a comma-separated list of items each a colon-separated string made up of a terminal type pattern (matched using `fnmatch(3)`) and a set of `name=value` entries.
> 
> For example, to set the `clear` `terminfo(5)` entry to `\e[H\e[2J` for all terminal types and the `dch1` entry to `\e[P` for the `rxvt` terminal type, the option could be set to the string:
> 
> `"*:clear=\e[H\e[2J,rxvt:dch1=\e[P"`
> 
> The terminal entry value is passed through `strunvis(3)` before interpretation.  The default value forcibly corrects the `colors` entry for terminals which support 256 colours:
> 
> `"*256col*:colors=256,xterm*:XT"`

So you should be able to simply use the glob approach to do something like this:

```
set-option -ga terminal-overrides ",*256col*:Tc"
```

if you prefer a whitelist-type approach, you could just enumerate known-working terminal types:

```
set-option -ga terminal-overrides ",xterm-256color:Tc,rxvt-unicode-256color:Tc"
```

**Comment by unphased** [comment]
Just a note in case anyones trying this. 

I was trying to get 24 bit color working for nested tmux. That is to say: `ssh -t tmux attach` INSIDE a tmux pane. 

It didnt work at first but I realized that if I add a corresponding rule for Tc in the overrides like so: 

`,xterm-256color-italic:Tc,tmux-256color-italic:Tc`

This works! Nested tmux 24 bit color.

**Comment by joshdick** [comment]
> So you should be able to simply use the glob approach to do something like this:
> `set-option -ga terminal-overrides ",*256col*:Tc"`

@jwhitley For some reason, when using this configuration:

```
set -g default-terminal "tmux" # use 'tmux' instead of 'screen-*' to enable italics support per tmux FAQ
set-option -ga terminal-overrides ",*256col*:Tc"
```

Most things seem to work as expected, but `tput colors` shows `8`, which appears to cause Vim 8.0 to display completely in black and white inside tmux. nvim v0.2.0 displays colors properly/as expected. Any idea why this is happening?

**Comment by jfelchner** [comment]
@ThomasAdam unfortunately people keep commenting on this issue for things that should be new issues. 
 Might be worth it to lock this conversation.  This feature is working great for most and any new comments should probably be new issues.

**Comment by rinetd** [comment]
```tmux
# Tmux开启24 bit color ！！！importent！！！ 其他方式都无效
set -g default-terminal "tmux-256color"
set -ga terminal-overrides ",*256col*:Tc"
```

**Comment by pickfire** [comment]
@rinetd 's/importent/important/'. 赞，我也是
